xds/src/main/java/io/grpc/xds/XdsLoadReportStore.java

```diff
@@ -0,0 +1,141 @@
+ */
+@ThreadSafe
+final class XdsLoadReportStore {
+
+  private static final ClientStreamTracer NOOP_CLIENT_STREAM_TRACER = new ClientStreamTracer() {
+  };
+  private final String clusterName;
+  private final ConcurrentMap<Locality, ClientLoadCounter> localityLoadCounters;
+  // Cluster level dropped request counts for each category specified in the DropOverload policy.
+  private final ConcurrentMap<String, AtomicLong> dropCounters;
+
+  XdsLoadReportStore(String clusterName) {
+    this(clusterName, new ConcurrentHashMap<Locality, ClientLoadCounter>(),
+        new ConcurrentHashMap<String, AtomicLong>());
+  }
+
+  @VisibleForTesting
+  XdsLoadReportStore(String clusterName,
+      ConcurrentMap<Locality, ClientLoadCounter> localityLoadCounters,
+      ConcurrentMap<String, AtomicLong> dropCounters) {
+    this.clusterName = checkNotNull(clusterName, "clusterName");
+    this.localityLoadCounters = checkNotNull(localityLoadCounters, "localityLoadCounters");
+    this.dropCounters = checkNotNull(dropCounters, "dropCounters");
+  }
+
+  /**
+   * Generates a {@link ClusterStats} containing load stats in locality granularity.
+   */
+  ClusterStats generateLoadReport(Duration interval) {
+    ClusterStats.Builder statsBuilder = ClusterStats.newBuilder().setClusterName(clusterName)
+        .setLoadReportInterval(interval);
+    for (Map.Entry<Locality, XdsClientLoadRecorder.ClientLoadCounter> entry : localityLoadCounters
+        .entrySet()) {
+      XdsClientLoadRecorder.ClientLoadSnapshot snapshot = entry.getValue().snapshot();
+      statsBuilder
+          .addUpstreamLocalityStats(UpstreamLocalityStats.newBuilder()
+              .setLocality(entry.getKey())
+              .setTotalSuccessfulRequests(snapshot.callsSucceed)
+              .setTotalErrorRequests(snapshot.callsFailed)
+              .setTotalRequestsInProgress(snapshot.callsInProgress));
+    }
+    for (Map.Entry<String, AtomicLong> entry : dropCounters.entrySet()) {
+      statsBuilder.addDroppedRequests(DroppedRequests.newBuilder()
+          .setCategory(entry.getKey())
+          .setDroppedCount(entry.getValue().get()));
+    }
+    return statsBuilder.build();
+  }
+
+  /**
+   * Create a {@link ClientLoadCounter} for the provided locality if not present in this {@link
+   * XdsLoadReportStore}. This method needs to be called at locality updates for newly assigned
+   * localities in balancer discovery responses.
+   */
+  void addLocality(Locality locality) {
+    localityLoadCounters.putIfAbsent(locality, new ClientLoadCounter());
+  }
+
+  /**
+   * Discard the {@link ClientLoadCounter} for the provided locality if owned by this {@link
+   * XdsLoadReportStore} to avoid map size growing infinitely. To be called at locality updates when
+   * the provided locality is no longer considered in balancer discovery response.s
+   */
+  void removeLocality(Locality locality) {
+    localityLoadCounters.remove(locality);
```

Comment:
You test code reminds me of long-running RPCs. If a locality is removed by xDS balancer while there are still running RPCs for it and they don't finish soon, there are still substantial load data for this locality to report. You may want to confirm with the design author about this, but my guess is that you don't want to remove it right away. If that's the case, a possible solution is to add a boolean "active" to `ClientLoadCounter`, so that `removeLocality()` only sets the boolean to false, `addLocality()` makes sure the boolean is true, and `generateLoadReport()` removes the entry if all counters are zero and the boolean is false.

Reply:
When calling `interceptPickResult` with an existing but inactive `ClientLoadCounter`, which means we want to start tracking a backend in the locality that is no longer exposed by the balancer, should we proceed?

Reply:
ACK.